components/org.wso2.carbon.crypto.impl/src/main/java/org/wso2/carbon/crypto/impl/DefaultCryptoService.java

```diff
@@ -501,6 +509,58 @@ public PrivateKey getPrivateKey(CryptoContext cryptoContext) throws CryptoExcept
         }
     }
 
+    /**
+     * Computes and returns the ciphertext of the given cleartext.
+     * @param cleartext                     The cleartext to be encrypted.
+     * @param algorithm                     The encryption / decryption algorithm
+     * @param javaSecurityAPIProvider       The Java Security API provider.
+     * @param returnSelfContainedCipherText Whether cipher text need to be self contained.
+     * @return
+     * @throws CryptoException
+     */
+    @Override
+    public byte[] encrypt(byte[] cleartext, String algorithm, String javaSecurityAPIProvider,
+                          boolean returnSelfContainedCipherText) throws CryptoException {
+
+        failIfInternalCryptoInputsAreNotValid(cleartext, algorithm, "'Internal Encryption'");
+        if (log.isDebugEnabled()) {
+
+            log.debug(String.format("Encrypting data using the algorithm '%s' and the Java Security API provider '%s'.",
+                    algorithm, javaSecurityAPIProvider));
+        }
+        if (areInternalCryptoProvidersAvailable()) {
+
+            InternalCryptoProvider mostSuitableInternalProvider = getMostSuitableInternalProvider();
+
+            if (log.isDebugEnabled()) {
+
+                log.debug(String.format("Internal providers are available. The most suitable provider is '%s'",
+                        mostSuitableInternalProvider.getClass().getCanonicalName()));
+            }
+            byte[] encryptedKey = mostSuitableInternalProvider
+                    .encrypt(cleartext, algorithm, javaSecurityAPIProvider);
+
+            if (returnSelfContainedCipherText) {
+                Certificate certificate = getCertificate(CryptoContext.buildEmptyContext(
+                        MultitenantConstants.SUPER_TENANT_ID, MultitenantConstants.SUPER_TENANT_DOMAIN_NAME));
+                try {
+                    encryptedKey = createSelfContainedCiphertext(encryptedKey, algorithm, certificate);
+                } catch (CertificateEncodingException e) {
+                    throw new CryptoException("Error while encoding certificate for self contained cipher text.", e);
+                } catch (NoSuchAlgorithmException e) {
+                    throw new CryptoException("Error while calculating thumb print for self contained cipher text.", e);
+                }
+            }
+            return encryptedKey;
+
+        } else {
+            String errorMessage = String.format("No internal crypto providers available. Correctly register " +
+                    "a service implementation of '%s' as an OSGi service", InternalCryptoProvider.class);
+            throw new CryptoException(errorMessage);
+        }
+
+    }
```

Comment:
Will we need to implement all this logic. Isn't there a method implementation that can be used to get the encrypted value, so that within this method it can be maid self contained

Reply:
I have refactored the code, such that I will be invoking a mostSuitableInternalProvider
                    .encrypt(cleartext, algorithm, javaSecurityAPIProvider, returnSelfContainedCipherText) in respective provider.
It is a new api.